node/cancellation-v2.md

```diff
@@ -0,0 +1,343 @@
+  // Timers and Activities are automatically cancelled when their containing scope is cancelled.
+  // Awaiting on a cancelled scope with throw the original CancellationError.
+  try {
+    await cancellationScope(async (scope) => {
+      sleep(1); // <-- Will be cancelled because it is attached to `scope`
+      scope.cancel();
+    });
+  } catch (e) {
+    if (e instanceof CancellationError) {
+      console.log('Exception was propagated 👍');
+    }
+    throw e; // <-- Fail the workflow
+  }
+}
+```
+
+#### Run multiple activities with a single deadline
+
+```ts
+import { withTimeout } from '@temporalio/workflow';
+import { httpGetJSON } from '@activities';
+
+export async function main(urls: string[], timeoutMs: number) {
+  // If timeout triggers before all activities complete
+  // the Workflow will fail with a CancellationError.
+  const results = await withTimeout(timeout, () => Promise.all(urls.map(httpGetJSON)));
+  // Do something with the results
+}
+```
+
+#### `CancellationScope`s can be shielded, preventing cancellation from propagating to created cancellables
+
+```ts
+import { cancellationScope } from '@temporalio/workflow';
+import { httpGetJSON } from '@activities';
+
+export async function main(url: string) {
+  // Prevent Activity from being cancelled and await completion.
+  // Note that the Workflow is completely oblivious to cancellation in this example.
+  const result = await cancellationScope(async (scope) => {
+    scope.shield = true;
+    return httpGetJSON(url);
+  });
+  return result;
+}
+```
+
+#### `Context.cancelled` may be awaited upon in order to make Workflow aware of cancellation while waiting on shielded scopes
+
+```ts
+import { CancellationError, cancellationScope } from '@temporalio/workflow';
+import { httpGetJSON } from '@activities';
+
+export async function main(url: string) {
+  let result: any = undefined;
+  try {
+    const promise = cancellationScope({ shield: true }, () => httpGetJSON(url));
+    result = await Promise.race([Context.cancelled, promise]);
+  } catch (err) {
+    if (!(e instanceof CancellationError)) {
+      throw err;
+    }
+    // Prevent Workflow from completing so Activity can complete
+    result = await promise;
+  }
+  return result;
+}
+```
+
+#### Handle Workflow cancellation by an external client while an Activity is running
+
+```ts
+import { CancellationError, cancellationScope } from '@temporalio/workflow';
+import { httpPostJSON, cleanup } from '@activities';
+
+export async function main(url: string, data: any): Promise<void> {
+  try {
+    await httpPostJSON(url, data);
+  } catch (e) {
+    if (e instanceof CancellationError) {
+      console.log('Workflow cancelled');
+      // Cleanup logic goes in a shielded scope here
+      // If we'd run cleanup outside of a shielded scope it would've been cancelled
+      // before being started because the Workflow's root scope is cancelled.
+      await cancellationScope({ shield: true }, () => cleanup(url));
```

Comment:
This is slightly funky. I get why, but it seems like a really easy mistake to make. Is there any way we could somehow check that `CancellationError` has already been thrown in this scope? Also a bit nasty but if it can work reliably saves users some debugging sessions.

Reply:
Those were my thoughts initially, @mfateev thought otherwise.
This behavior also matches the cancel token example in the [trio blog post](https://vorpus.org/blog/timeouts-and-cancellation-for-humans/#cancel-tokens-are-level-triggered-and-can-be-scoped-to-match-your-program-s-needs).

I also feel like devs will likely get this wrong though, and I'm not sure this behavior is suitable for activities.
Let's go over this again with @mfateev.